### PR TITLE
Add custom cache layer for session loading

### PR DESCRIPTION
**Pull request by fendor**
Achieves this by adding a HashMap from NormalizedFilePath
to its respective `hie.yaml` location.
This works nicely with the existing implementation, but
increases the memory usage (and I am not sure whether by a considerable amount).

This change is motivated by https://github.com/mpickering/hie-bios/pull/264
where we set the build-dir for cabal based projects to some config
directory in order to avoid recompilation issues whenever users invoked
`cabal build`.
However, this uncovered an issue in the existing code-base, as HLS will
ask for the compilation options for generated modules, such as
`Paths_*`, by looking for the responsible `hie.yaml` file and using an
internal cache for the options. This used to be fine, until the
aforementioned hie-bios change, since now `Paths_*` module will be in
some `~/.cache/hie-bios/...` directory, which has no responsible
`hie.yaml` which is why we see error messages such: `No hie.yaml found,
use implicit hie-bios`.

I *think* this error message was only shown when an explicit `hie.yaml` is used.

I hope the motivation and explanation was understandable :) (and sound, and correct)

The implementation is more a POC, to show that it really fixes the problem.

You can give it a try and see the changes by checking out https://github.com/fendor/haskell-language-server/tree/cabal-dist-dir-tests and using it on `exe:hie-bios`. Compare it to the output HLS gives when reverting the last commit to that branch.

cc @wz1000

### CODE REVIEW

**Comment by jneira** [comment]
Do we have some test around `Path_`'s module that were broken with that version of hie-bios?

**Comment by wz1000** [comment]
Wouldn't it be simpler to make hie-bios detect and return the correct cradle location for files in `~/.cache/hie-bios/...`?

**Comment by fendor** [comment]
@wz1000 The invoked function is `findCradle :: FilePath -> IO (Maybe FilePath)`, where the first parameter is `~/.cache/.../Paths_*`, so we actually don't have the required information in that function, afaict.

Also, I think this is rather a HLS bug than hie-bios' bug, since hie-bios lists all modules and options (including Paths_* module), HLS should be able to work with it. Thus, the patch here.

Another possible implementation approach might be to utilise the Shake graph in some way.

**Comment by wz1000** [comment]
We can write the path to the original hie.yaml file/root  in something like `~/.cache/hie-bios/<builddir>/root.path`

I'm also concerned about the behavior with other kinds of preprocessed haskell files, like happy parsers, hsc files etc. I think even they will end up being written in the builddir.

**Comment by fendor** [comment]
Seems kind of hacky to me, when HLS should have all the required information already (e.g. `-i`, `-I`,...) and figure that out itself.

**Comment by jneira** [comment]
I marked it as draft in github following the description, to no merge it accidentally

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 1]
```diff
@@ -329,7 +333,9 @@ loadSessionWithOptions SessionLoadingOptions{..} dir = do
                let res = (map (renderCradleError ncfp) err, Nothing)
                modifyVar_ fileToFlags $ \var -> do
                  pure $ Map.insertWith HM.union hieYaml (HM.singleton ncfp (res, dep_info)) var
-               return (res, maybe [] pure hieYaml ++ concatMap cradleErrorDependencies err)
+               modifyVar_ filesMap $ \var -> do
+                 pure $ HM.insert ncfp hieYaml var
+               return (res,[])
```
Comment (pepeiborra) [inline ghcide/session-loader/Development/IDE/Session.hs:338]: what's the reason to throw away the cradle error deps here?

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 2]
```diff
@@ -358,9 +364,11 @@ loadSessionWithOptions SessionLoadingOptions{..} dir = do
     -- before attempting to do so.
     let getOptions :: FilePath -> IO (IdeResult HscEnvEq, [FilePath])
         getOptions file = do
+            ncfp <- toNormalizedFilePath' <$> canonicalizePath file
+            cachedHieYamlLocation <- HM.lookup ncfp <$> readVar filesMap
             hieYaml <- cradleLoc file
-            sessionOpts (hieYaml, file) `catch` \e ->
-                return (([renderPackageSetupException file e], Nothing), maybe [] pure hieYaml)
+            sessionOpts (join cachedHieYamlLocation <|> hieYaml, file) `catch` \e ->
+                return (([renderPackageSetupException file e], Nothing),[])
```
Comment (pepeiborra) [inline ghcide/session-loader/Development/IDE/Session.hs:371]: same here, why throw away the error deps? Should be the selected cradle

**Comment by pepeiborra** [comment]
I'm not too worried by the increase in memory usage, did you check the benchmarks?

**Comment by pepeiborra** [comment]
I think that we should push ahead with this, unless there is a better solution available. Waiting for Cabal is just mad

File: ghcide/session-loader/Development/IDE/Session.hs (hunk 1 above)
Comment (fendor) [reply ghcide/session-loader/Development/IDE/Session.hs:338]: Accident, sorry!

**Comment by fendor** [comment]
I will check the benchmarks and address the feedback, thank you!

**Comment by fendor** [comment]
Having a bit of trouble trying to make sense of it. 


hover

getDefinition_after_edit

code_actions


code_actions_after_edit

**Comment by wz1000** [comment]
Those are just artifacts. The graphs are identical for all purposes.

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 3]
```diff
@@ -329,6 +337,8 @@ loadSessionWithOptions SessionLoadingOptions{..} dir = do
                let res = (map (renderCradleError ncfp) err, Nothing)
                modifyVar_ fileToFlags $ \var -> do
                  pure $ Map.insertWith HM.union hieYaml (HM.singleton ncfp (res, dep_info)) var
+               modifyVar_ filesMap $ \var -> do
+                 pure $ HM.insert ncfp hieYaml var
```
Comment (pepeiborra) [inline ghcide/session-loader/Development/IDE/Session.hs:341]:
```suggestion
                 evaluate $ HM.insert ncfp hieYaml var
```

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 4]
```diff
@@ -271,6 +277,8 @@ loadSessionWithOptions SessionLoadingOptions{..} dir = do
 
           modifyVar_ fileToFlags $ \var -> do
               pure $ Map.insert hieYaml (HM.fromList (concatMap toFlagsMap all_targets)) var
+          modifyVar_ filesMap $ \var -> do
+              pure $ HM.union var (HM.fromList (zip (map fst $ concatMap toFlagsMap all_targets) (repeat hieYaml)))
```
Comment (pepeiborra) [inline ghcide/session-loader/Development/IDE/Session.hs:281]:
```suggestion
              evaluate $ HM.union var (HM.fromList (zip (map fst $ concatMap toFlagsMap all_targets) (repeat hieYaml)))
```

**Review by pepeiborra — APPROVED** [review]
I think this is ready to merge

**Comment by fendor** [comment]
I will make a release for hie-bios with the mentioned change, and we will see if there is any fallout :)

**Comment by wz1000** [comment]
@fendor I think https://github.com/mpickering/hie-bios/pull/265 is also important in order to actually make ghcide startup faster.

**Comment by fendor** [comment]
@wz1000 True, but it already improves the developer flow by not forcing the user to re-configure whenever they use cabal.

**Pull request merged**